services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java

```diff
@@ -60,34 +64,58 @@
+        if (authSession == null) {
+            // we have no authSession so we cannot create an authenticator list.
+            return Collections.emptyList();
+        }
 
-            String topFlowId = getFlowIdOfTheHighestUsefulFlow(processor, model);
+        Map<String, AuthenticationExecutionModel> typeAuthExecMap = new HashMap<>();
+        List<AuthenticationExecutionModel> nonCredentialExecutions = new ArrayList<>();
 
-            if (topFlowId == null) {
-                addSimpleAuthenticationExecution(processor, model, typeAuthExecMap, nonCredentialExecutions);
-            } else {
-                addAllExecutionsFromSubflow(processor, topFlowId, typeAuthExecMap, nonCredentialExecutions);
-            }
+        String topFlowId = getFlowIdOfTheHighestUsefulFlow(processor, model);
+
+        if (topFlowId == null) {
+            addSimpleAuthenticationExecution(processor, model, typeAuthExecMap, nonCredentialExecutions);
+        } else {
+            addAllExecutionsFromSubflow(processor, topFlowId, typeAuthExecMap, nonCredentialExecutions);
+        }
+
+        List<AuthenticationSelectionOption> authenticationSelectionList = new ArrayList<>();
+        KeycloakSession session = processor.getSession();
+        UserModel authenticatedUser = authSession.getAuthenticatedUser();
+        // add authenticator options for the given user if present
+        if (authenticatedUser != null) {
+
+            RealmModel realm = processor.getRealm();
 
             //add credential authenticators in order
-            if (processor.getAuthenticationSession().getAuthenticatedUser() != null) {
-                authenticationSelectionList = processor.getSession().userCredentialManager()
-                        .getStoredCredentialsStream(processor.getRealm(), processor.getAuthenticationSession().getAuthenticatedUser())
-                        .filter(credential -> typeAuthExecMap.containsKey(credential.getType()))
-                        .map(CredentialModel::getType)
-                        .distinct()
-                        .map(credentialType -> new AuthenticationSelectionOption(processor.getSession(), typeAuthExecMap.get(credentialType)))
-                        .collect(Collectors.toList());
+            authenticationSelectionList.addAll(session.userCredentialManager()
+                    .getStoredCredentialsStream(realm, authenticatedUser)
+                    .filter(credential -> typeAuthExecMap.containsKey(credential.getType()))
+                    .map(CredentialModel::getType)
+                    .distinct()
+                    .map(credentialType -> new AuthenticationSelectionOption(session, typeAuthExecMap.get(credentialType)))
+                    .collect(Collectors.toList()));
+
+            //add authenticators that require a current user which are also configured for that user
+            for (AuthenticationExecutionModel exec : nonCredentialExecutions) {
```

Comment:
If authenticated user is not null, this list of executions is fully iterated twice. In this case, what about to reduce these executions or completely avoid the second iteration over whole list? The condition for the `requiresUser` is evaluated twice. I know the list would change and potentially break the process of executing actions for the `nonCredentialExecutions` list in the future, but at this moment, there's no other action for the list. Is there any deeper intention why to use this approach?